src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+        self._close()
+        if self.ad['verbose_logging']:
+            self.logger.debug(f'Query for nETBIOSName from LDAP returned: [{netbios_name}]')
+        return netbios_name
+
+    def locate_site(self):
+        """
+        In Windows environment, this is discovered via CLDAP query for closest DC. We
+        can't do this, and so we have to rely on comparing our network configuration with
+        site and subnet information obtained through LDAP queries. 
+        """
+        if not self._handle:
+            self._open()
+        ipv4_site = None
+        ipv6_site = None
+        self._initialize_naming_context() 
+        subnets = self._get_subnets()
+        for nic in self.interfaces:
+            for alias in nic['aliases']:
+                if alias['type'] == 'INET':
+                    if ipv4_site is not None:
+                        continue
+                    ipv4_addr_obj = ipaddr.IPAddress(alias['address'], version=4)
+                    for subnet in subnets['ipv4_subnet_info']:
+                        if ipv4_addr_obj in subnet['network']:
+                            sinfo = self._get_sites(distinguishedname=subnet['site_dn'])[0]
+                            if sinfo and len(sinfo) > 1:
+                                ipv4_site = sinfo[1]['cn'][0].decode()
+                                break
+                        
+                if alias['type'] == 'INET6':
+                    if ipv6_site is not None:
+                        continue
+                    ipv6_addr_obj = ipaddr.IPAddress(alias['address'], version=6)
+                    for subnet in subnets['ipv6_subnet_info']:
+                        if ipv6_addr_obj in subnet['network']:
+                            sinfo = self._get_sites(distinguishedname=subnet['site_dn'])[0]
+                            if sinfo and len(sinfo) > 1:
+                                ipv6_site = sinfo[1]['cn'][0].decode()
+                                break
+                        
+        if ipv4_site and ipv6_site and ipv4_site == ipv6_site:
+            return ipv4_site
+
+        if ipv4_site:
+            return ipv4_site
+
+        if not ipv4_site and ipv6_site:
+            return ipv6_site
+
+        return None 
+
+class ActiveDirectoryService(ConfigService):
+    class Config:
+        service = "activedirectory"
+        datastore = 'directoryservice.activedirectory'
+        datastore_extend = "activedirectory.ad_extend"
+        datastore_prefix = "ad_"
+
+    @private
+    async def ad_extend(self, ad):
+        return ad 
+
+    @private
+    async def ad_compress(self, ad):
+        if ad['kerberos_realm']:
+            ad['kerberos_realm'] = ad['kerberos_realm']['id']
+
+        return ad 
+
+    @accepts(Dict(
+        'activedirectory_update',
+        Str('domainname', required=True),
+        Str('bindname'),
+        Str('bindpw'),
+        Str('ssl', default='off', enum=['off', 'on', 'start_tls']),
+        Dict('certificate'),
```

Comment:
Why certificate is a dict? Do you mean Int?

Reply:
Yes.